spec.html

```diff
@@ -39920,8 +40074,8 @@ <h1>Properties of the %TypedArray% Prototype Object</h1>
         <h1>%TypedArray%.prototype.at ( _index_ )</h1>
         <emu-alg>
           1. Let _O_ be the *this* value.
-          1. Perform ? ValidateTypedArray(_O_).
-          1. Let _len_ be _O_.[[ArrayLength]].
+          1. Let _iieoRecord_ be ? ValidateTypedArray(_O_, ~SeqCst~).
```

Comment:
It may make sense to add a note somewhere to explain that all `%TypedArray%.prototype` synchronise the length, even when doing only bounds checking.

Given an initially zero length growable SharedArrayBuffer `sab` when a concurrent thread (atomically) grows the length to 4 bytes and writes `123` at offset zero. The following results should be possible:

| Operation | Condition | Result |
|---|---|---|
| `new Int32Array(sab)[0]` | New length not read | `undefined` for out-of-bounds |
| `new Int32Array(sab)[0]` | New content not read | `0` |
| `new Int32Array(sab)[0]` | New length and content read | `123` |
| `new Int32Array(sab).at(0)` | New content not read | `0` |
| `new Int32Array(sab).at(0)` | New content read | `123` |
| `Atomics.load(new Int32Array(sab), 0)` | New length not read | `RangeError` for out-of-bounds |
| `Atomics.load(new Int32Array(sab), 0)` | New length read | `123` |

When the operations are performed in exactly this order:

| Thread 1  | Thread 2 |
| --- | --- |
| `ta = new Int32Array(sab)` | - |
| - | `sab.grow(4)` |
| - | `Atomics.store(new Int32Array(sab), 0, 123)` |
| `ta[0]` or `ta.at(0)` or `Atomics.load(ta, 0)` | - | 


IIUC [get %TypedArray%.prototype.length](https://tc39.es/ecma262/#sec-get-%typedarray%.prototype.length) was repurposed as a synchronisation point for reading the (byte-)length. And `%TypedArray%.prototype` methods generally act as if `get %TypedArray%.prototype.length` is called and therefore all `%TypedArray%.prototype` methods now read the length with _SeqCst_. This also ensures `%TypedArray%.prototype` methods and user-implemented functions, which call `get %TypedArray%.prototype.length`, have the same semantics.

This isn't explained anywhere in the actual spec text though, which is likely to cause confusion in the future for people how don't know all these background information.

Reply:
Good call. I'll try to capture this in the "for implementers" section. The design principle was "implicit loads of the length are not synchronizing; explicit loads of the length are synchronizing".

And from that we derive the following positions:

- Explicitly accessing the `.length` or `.byteLength` properties **are synchronizing**.
- Built-in methods need to check whether TypedArrays are out-of-bounds or detached. This conceptually feels like an explicit access of the length (and may indeed be an explicit access of the length in self-hosted implementations), and so such loads of the length **are synchronizing**.
- Bounds checking for element access (i.e. integer-indexed property access, `ta[idx]`) is **not synchronizing**, because it feels implicit.

Reply:
Specifically, the "implicit loads are not synchronizing" is desirable because it is desirable to allow the optimization of hoisting bounds checks of TypedArray accesses out of loop bodies. If it were specced that all element access must have an SC load of the length, then that optimization would no longer be allowed.

Reply:
I wrote up some guidelines in https://github.com/tc39/ecma262/pull/3116/commits/30a73595bb5d6a937ca4a50524b8b71004565046, PTAL